### DISCUSSION TITLE
extend from a class with json fields but raise error: <class 'list'> has no matching SQLAlchemy type

### DISCUSSION THREAD

**Discussion by Ka2e** [fastapi/sqlmodel#1051] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel import SQLModel, Field, Column
from sqlalchemy import JSON


class User(SQLModel, table=True):
    __tablename__ = "t_user"
    id: int = Field(primary_key=True)
    num_list: list = Field(default=[], sa_column=Column(JSON))


class UserInfo(User, table=False):
    info: str | None = Field(default="")
```

### Description

I have some class extends from sqlmodel class with json fields, but it raise error:
```python
\venv\Lib\site-packages\sqlmodel\main.py", line 697, in get_sqlalchemy_type
    raise ValueError(f"{type_} has no matching SQLAlchemy type")
ValueError: <class 'list'> has no matching SQLAlchemy type
```

### Operating System

Windows

### SQLModel Version

0.0.21

### Python Version

3.11

**Comment by vlavorini** [comment]
Hello, same issue here, even with a simpler model like:

```python
from sqlmodel import create_engine, SQLModel, Session, Field


class Example(SQLModel, table=True):
    file_path: str = Field(default=None, primary_key=True)
    person_number: int
    person_list: list[str]


engine = create_engine("postgresql://....", echo=True)
SQLModel.metadata.create_all(engine)

to_rec = Example(file_path="asd",
                        person_number=2,
                        person_list=['a', 'b'])
with Session(engine) as session:
    session.add(to_rec)
    session.commit()
```

Running with 
SQLModel 0.0.21
sqlalchemy 2.0.32
Pydantic 2.5.2
python 3.10 on Linux

**Comment by KunxiSun** [comment]
Hello! Glad I can help.@Ka2e @vlavorini 
Do that 👇🏻

```python
from sqlmodel import (
    SQLModel,
    Field,
    Column,
    JSON,
)


class User(SQLModel, table=True):
    __tablename__ = "t_user"
    id: int = Field(primary_key=True)
    num_list: list = Field(default=[], sa_type=JSON)
```

The magic here is to use the field `sa_type`.

**Reply by vlavorini** [reply]
Indeed it works!

For completeness, I changed it a bit to exploit the PostgreSQL backend:

```python
num_list: list[str] = Field(default=None, sa_type=ARRAY(Text))
```

**Reply by Ka2e** [reply]
Thanks for your answer.
If we only use class `User`，it works well.
But if we declare a class `UserInfo` extend from class `User`, the error still exist:
```python
from sqlmodel import (
    SQLModel,
    Field,
    JSON,
)


class User(SQLModel, table=True):
    __tablename__ = "t_user"
    id: int = Field(primary_key=True)
    num_list: list = Field(default=[], sa_type=JSON)


class UserInfo(User, table=False):
    info: str | None = Field(default="")
```
ValueError: <class 'list'> has no matching SQLAlchemy type

**Reply by KunxiSun** [reply, marked as answer]
> Thanks for your answer. If we only use class `User`，it works well. But if we declare a class `UserInfo` extend from class `User`, the error still exist:
> 
> ```python
> from sqlmodel import (
>     SQLModel,
>     Field,
>     JSON,
> )
> 
> 
> class User(SQLModel, table=True):
>     __tablename__ = "t_user"
>     id: int = Field(primary_key=True)
>     num_list: list = Field(default=[], sa_type=JSON)
> 
> 
> class UserInfo(User, table=False):
>     info: str | None = Field(default="")
> ```
> 
> ValueError: <class 'list'> has no matching SQLAlchemy type

Oh, I missed your point. The output shows it is an error of inheriting from the database model.

I did meet this bug before, then I solve it by not inherit from the database model.

I split the model into different functions. I have:
`BaseModel`: has all shared field
`ModelIn`: responsible for the input data
`Model`: responsible for the database model, including foreign key, relationships 
`ModelPublic`: responsible for the data output

So in your case, it will be 👇🏻
```python
from somewhere import id_factory
from sqlmodel import (
    SQLModel,
    Field,
    JSON,
)

class BaseUser(SQLModel):
    """Have all the shared fields of Input, database model, and output"""
    id: int = Field(primary_key=True, default=id_factory)
    num_list: list = Field(default=[], sa_type=JSON)

class UserIn(BaseUser):
    """Accept responsible for the input from frontend"""

class User(BaseUser, table=True):
    """Database model"""

class UserPublic(BaseUser):
    """Output for frontend"""
    info: str | None = Field(default="")

```
